File: packages/go/qorechain/pqcdx/pqcdx.go
// Package pqcdx provides a high-level, quantum-safe developer-experience (DX)
// helper for QoreChain's x/pqc module. It wraps the three things an app developer
// actually wants when adopting post-quantum signatures:
//
//   - reading whether an address has a PQC key registered (IsPQCRegistered /
//     GetPQCStatus, via the qor_getPQCKeyStatus JSON-RPC method);
//   - idempotently registering the signer's PQC key if it is missing
//     (EnsurePQCRegistered, which builds and broadcasts a MsgRegisterPQCKeyV2 only
//     when needed); and
//   - migrating an account to hybrid (classical + ML-DSA-87) signing once the key
//     is registered (MigrateToHybrid / MigratePQCKey).
//
// The helper mirrors the SDK's module-helper style (see the crossvm helper): a
// caller-supplied Signer carries the explicit account number, sequence, chain id,
// and fee — the Go SDK never simulates or auto-fetches these — and the high-level
// methods build, sign, and broadcast using that context.
//
// On-chain mapping: MsgRegisterPQCKeyV2 { sender, public_key, algorithm_id,
// ecdsa_pubkey, key_type } at /qorechain.pqc.v1.MsgRegisterPQCKeyV2 (the chain's
// classical-exempt bootstrap path), and MsgMigratePQCKey at
// /qorechain.pqc.v1.MsgMigratePQCKey. Registration status is read via
// qor_getPQCKeyStatus.
package pqcdx

import (
	"encoding/json"
	"fmt"

	sdk "github.com/cosmos/cosmos-sdk/types"

	"github.com/qorechain/qorechain-sdk/packages/go/qorechain/accounts"
	"github.com/qorechain/qorechain-sdk/packages/go/qorechain/messages"
	"github.com/qorechain/qorechain-sdk/packages/go/qorechain/pqc"
	pqcv1 "github.com/qorechain/qorechain-sdk/packages/go/qorechain/proto/qorechain/pqc/v1"
	"github.com/qorechain/qorechain-sdk/packages/go/qorechain/tx"
)

// Message type URLs for the x/pqc messages this helper builds.
const (
	// MsgRegisterPQCKeyTypeURL is the type URL of the legacy (v1) PQC key
	// registration message.
	MsgRegisterPQCKeyTypeURL = "/qorechain.pqc.v1.MsgRegisterPQCKey"
	// MsgRegisterPQCKeyV2TypeURL is the type URL of the current PQC key
	// registration message (explicit algorithm_id; the chain's classical-exempt
	// bootstrap path). This is what the helpers in this package broadcast.
	MsgRegisterPQCKeyV2TypeURL = "/qorechain.pqc.v1.MsgRegisterPQCKeyV2"
	// MsgMigratePQCKeyTypeURL is the type URL of the PQC key migration message.
	MsgMigratePQCKeyTypeURL = "/qorechain.pqc.v1.MsgMigratePQCKey"
)

// DefaultKeyType is the key_type recorded on MsgRegisterPQCKeyV2 when the
// caller does not set RegisterOptions.KeyType. "hybrid" (classical + PQC) is
// what the chain's bootstrap registration path expects.
const DefaultKeyType = "hybrid"

// QorRPC is the subset of query.QorClient used to read PQC key status. The
// concrete *query.QorClient satisfies it.
type QorRPC interface {
	GetPQCKeyStatus(address string) (json.RawMessage, error)
}

// Status is the decoded result of qor_getPQCKeyStatus for an address.
type Status struct {
	// Registered reports whether the address has a PQC key on chain.
	Registered bool
	// AlgorithmID is the registered algorithm identifier (0 when unregistered;
	// 1 = Dilithium-5 / ML-DSA-87).
	AlgorithmID uint8
	// Pubkey is the registered PQC public key bytes (nil when unregistered).
	Pubkey []byte
}

// statusWire tolerantly decodes the qor_getPQCKeyStatus response. The chain
// reports the field under one of a few common names, and the public key may
// arrive as base64 (Go []byte JSON) or 0x-hex; both are accepted.
type statusWire struct {
	Registered   *bool  `json:"registered"`
	IsRegistered *bool  `json:"is_registered"`
	HasKey       *bool  `json:"has_key"`
	AlgorithmID  *uint8 `json:"algorithm_id"`
	AlgID        *uint8 `json:"algorithm"`
	Pubkey       []byte `json:"pubkey"`
	PublicKey    []byte `json:"public_key"`
}

func (w statusWire) toStatus() Status {
	s := Status{}
	switch {
	case w.Registered != nil:
		s.Registered = *w.Registered
	case w.IsRegistered != nil:
		s.Registered = *w.IsRegistered
	case w.HasKey != nil:
		s.Registered = *w.HasKey
	}
	switch {
	case w.AlgorithmID != nil:
		s.AlgorithmID = *w.AlgorithmID
	case w.AlgID != nil:
		s.AlgorithmID = *w.AlgID
	}
	switch {
	case len(w.Pubkey) > 0:
		s.Pubkey = w.Pubkey
	case len(w.PublicKey) > 0:
		s.Pubkey = w.PublicKey
	}
	// A non-empty key implies registration even if the boolean flag is absent.
	if !s.Registered && len(s.Pubkey) > 0 {
		s.Registered = true
	}
	return s
}

// GetPQCStatus reads the PQC key registration status for an address via
// qor_getPQCKeyStatus and decodes it into a Status.
func GetPQCStatus(client QorRPC, address string) (Status, error) {
	if client == nil {
		return Status{}, fmt.Errorf("pqcdx: GetPQCStatus requires a QorRPC client")
	}
	raw, err := client.GetPQCKeyStatus(address)
	if err != nil {
		return Status{}, err
	}
	var wire statusWire
	if err := json.Unmarshal(raw, &wire); err != nil {
		return Status{}, fmt.Errorf("pqcdx: decode qor_getPQCKeyStatus: %w", err)
	}
	return wire.toStatus(), nil
}

// IsPQCRegistered reports whether an address has a PQC key registered, via
// qor_getPQCKeyStatus.
func IsPQCRegistered(client QorRPC, address string) (bool, error) {
	s, err := GetPQCStatus(client, address)
	if err != nil {
		return false, err
	}
	return s.Registered, nil
}

// Signer carries the signing context EnsurePQCRegistered / MigrateToHybrid need
// to build, sign, and broadcast a transaction. It mirrors the crossvm helper's
// Signer: the account number / sequence / chain id / fee are supplied explicitly
// (as everywhere else in tx.*). The sender address is taken from Account.Address.
type Signer struct {
	// Account is the native secp256k1 signer. Its PublicKey (33-byte compressed
	// secp256k1 key) is recorded as the message's ecdsa_pubkey.
	Account accounts.Secp256k1Account
	// PQCKeypair is the signer's ML-DSA-87 (Dilithium-5) keypair. Its PublicKey
	// is recorded as the message's dilithium_pubkey, and the secret key is used
	// for hybrid signing in MigrateToHybrid.
	PQCKeypair pqc.Keypair
	// ChainID is the chain id (e.g. "qorechain-vladi").
	ChainID string
	// RestURL is the REST (LCD) base URL used to broadcast.
	RestURL string
	// AccountNumber is the signer's on-chain account number.
	AccountNumber uint64
	// Sequence is the signer's current account sequence (nonce).
	Sequence uint64
	// Fee is the explicit fee to pay.
	Fee tx.Fee
	// Mode is the broadcast mode (defaults to tx.BroadcastSync).
	Mode tx.BroadcastMode
	// Wait configures post-broadcast polling (zero values use sensible defaults).
	Wait tx.WaitOptions
}

// Client is the high-level PQC DX helper bound to a Signer and a status reader.
type Client struct {
	signer Signer
	qor    QorRPC
}

// Options configures New.
type Options struct {
	// Qor is the qor_* JSON-RPC client used to read registration status
	// (qor_getPQCKeyStatus). Required for EnsurePQCRegistered's idempotency check.
	Qor QorRPC
}

// New creates a PQC DX helper bound to a Signer.
func New(signer Signer, opts Options) *Client {
	return &Client{signer: signer, qor: opts.Qor}
}

// Address returns the signer's native address.
func (c *Client) Address() string { return c.signer.Account.Address }

// Status reads the signer's own PQC registration status.
func (c *Client) Status() (Status, error) {
	return GetPQCStatus(c.qor, c.signer.Account.Address)
}

// RegisterOptions tune EnsurePQCRegistered.
type RegisterOptions struct {
	// KeyType is the key_type recorded on MsgRegisterPQCKeyV2. Empty uses
	// DefaultKeyType.
	KeyType string
	// Force skips the qor_getPQCKeyStatus pre-check and always broadcasts the
	// registration message. Use when status is already known to be unregistered.
	Force bool
}

// EnsureResult reports the outcome of EnsurePQCRegistered.
type EnsureResult struct {
	// AlreadyRegistered is true when the address already had a PQC key and no
	// transaction was broadcast.
	AlreadyRegistered bool
	// TxHash is the registration tx hash. Empty when AlreadyRegistered is true.
	TxHash string
}

// registerMessage builds the MsgRegisterPQCKeyV2 for the signer.
func (c *Client) registerMessage(keyType string) (*pqcv1.MsgRegisterPQCKeyV2, error) {
	if len(c.signer.PQCKeypair.PublicKey) == 0 {
		return nil, fmt.Errorf("pqcdx: Signer.PQCKeypair.PublicKey is required to register")
	}
	if len(c.signer.Account.PublicKey) == 0 {
		return nil, fmt.Errorf("pqcdx: Signer.Account.PublicKey is required to register")
	}
	if keyType == "" {
		keyType = DefaultKeyType
	}
	return messages.Pqc.RegisterKeyV2(
		c.signer.Account.Address,
		c.signer.PQCKeypair.PublicKey,
		pqcv1.AlgorithmID(pqc.AlgorithmDilithium5),
		c.signer.Account.PublicKey,
		keyType,
	), nil
}

// EnsurePQCRegistered ensures the signer's PQC key is registered on chain. It is
// idempotent: unless RegisterOptions.Force is set, it first checks
// qor_getPQCKeyStatus and returns AlreadyRegistered without broadcasting when a
// key is present; otherwise it builds and broadcasts a MsgRegisterPQCKeyV2
// carrying the signer's Dilithium and ECDSA public keys.
func (c *Client) EnsurePQCRegistered(opts RegisterOptions) (EnsureResult, error) {
	if !opts.Force {
		s, err := c.Status()
		if err != nil {
			return EnsureResult{}, err
		}
		if s.Registered {
			return EnsureResult{AlreadyRegistered: true}, nil
		}
	}
	msg, err := c.registerMessage(opts.KeyType)
	if err != nil {
		return EnsureResult{}, err
	}
	res, err := c.broadcast([]sdk.Msg{msg})
	if err != nil {
		return EnsureResult{}, err
	}
	return EnsureResult{TxHash: res.TxHash}, nil
}

// BuildRegister builds and signs the MsgRegisterPQCKeyV2 WITHOUT broadcasting,
// returning the BuiltTx. It does not check current status.
func (c *Client) BuildRegister(opts RegisterOptions) (*tx.BuiltTx, error) {
	msg, err := c.registerMessage(opts.KeyType)
	if err != nil {
		return nil, err
	}
	return c.build([]sdk.Msg{msg})
}

// MigrateOptions tune MigratePQCKey.
type MigrateOptions struct {
	// OldPublicKey is the previously registered PQC public key.
	OldPublicKey []byte
	// NewPublicKey is the replacement PQC public key.
	NewPublicKey []byte
	// NewAlgorithmID is the replacement algorithm identifier (e.g.
	// pqc.AlgorithmDilithium5).
	NewAlgorithmID uint8
	// OldSignature proves control of the old key (signature over the migration
	// challenge with the old secret key).
	OldSignature []byte
	// NewSignature proves control of the new key (signature over the migration
	// challenge with the new secret key).
	NewSignature []byte
}

// migrateMessage builds the MsgMigratePQCKey for the signer.
func (c *Client) migrateMessage(o MigrateOptions) (*pqcv1.MsgMigratePQCKey, error) {
	if len(o.NewPublicKey) == 0 {
		return nil, fmt.Errorf("pqcdx: MigrateOptions.NewPublicKey is required")
	}
	if len(o.OldSignature) == 0 || len(o.NewSignature) == 0 {
		return nil, fmt.Errorf("pqcdx: MigrateOptions requires OldSignature and NewSignature")
	}
	return messages.Pqc.MigrateKey(
		c.signer.Account.Address,
		o.OldPublicKey,
		o.NewPublicKey,
		pqcv1.AlgorithmID(o.NewAlgorithmID),
		o.OldSignature,
		o.NewSignature,
	), nil
}

// MigratePQCKey builds, signs, and broadcasts a MsgMigratePQCKey, rotating the
// account's registered PQC key to a new key (and optionally a new algorithm).
func (c *Client) MigratePQCKey(o MigrateOptions) (*tx.TxResult, error) {
	msg, err := c.migrateMessage(o)
	if err != nil {
		return nil, err
	}
	return c.broadcast([]sdk.Msg{msg})
}

// BuildMigratePQCKey builds and signs a MsgMigratePQCKey WITHOUT broadcasting.
func (c *Client) BuildMigratePQCKey(o MigrateOptions) (*tx.BuiltTx, error) {
	msg, err := c.migrateMessage(o)
	if err != nil {
		return nil, err
	}
	return c.build([]sdk.Msg{msg})
}

// MigrateOptionsHybrid tune MigrateToHybrid.
type MigrateToHybridOptions struct {
	// Register controls the pre-flight registration: when set, MigrateToHybrid
	// runs EnsurePQCRegistered with these options first.
	Register RegisterOptions
	// IncludePQCPublicKey embeds the signer's Dilithium public key in the hybrid
	// signature extension (B-mode). Leave false for the B0 contract (no embedded
	// public key) when the chain already knows the key.
	IncludePQCPublicKey bool
}

// MigrateToHybridResult reports the outcome of MigrateToHybrid.
type MigrateToHybridResult struct {
	// AlreadyRegistered mirrors EnsureResult.AlreadyRegistered from the pre-flight
	// registration step.
	AlreadyRegistered bool
	// RegisterTxHash is the registration tx hash (empty when AlreadyRegistered).
	RegisterTxHash string
	// Result is the confirmed hybrid transaction result.
	Result *tx.TxResult
}

// MigrateToHybrid is the one-call "go quantum-safe" path: it ensures the signer's
// PQC key is registered, then builds and broadcasts the given messages as a
// hybrid (classical + ML-DSA-87) transaction using the signer's PQC keypair.
//
// The messages are any module messages (e.g. a bank send from the composers);
// they are sent under the hybrid signature so the account's first hybrid tx and
// its registration can be sequenced without hand-wiring tx.BuildHybridMessages.
func (c *Client) MigrateToHybrid(msgs []sdk.Msg, opts MigrateToHybridOptions) (*MigrateToHybridResult, error) {
	if len(msgs) == 0 {
		return nil, fmt.Errorf("pqcdx: MigrateToHybrid requires at least one message")
	}
	if len(c.signer.PQCKeypair.SecretKey) == 0 {
		return nil, fmt.Errorf("pqcdx: Signer.PQCKeypair.SecretKey is required for hybrid signing")
	}
	ensure, err := c.EnsurePQCRegistered(opts.Register)
	if err != nil {
		return nil, err
	}
	// When a registration tx was broadcast in the same flow, advance the local
	// sequence so the subsequent hybrid tx signs over the next nonce.
	sequence := c.signer.Sequence
	if !ensure.AlreadyRegistered {
		sequence++
	}
	built, err := tx.BuildHybridMessages(tx.BuildHybridMessagesParams{
		Account:             c.signer.Account,
		PQCKeypair:          c.signer.PQCKeypair,
		Messages:            msgs,
		Fee:                 c.signer.Fee,
		ChainID:             c.signer.ChainID,
		AccountNumber:       c.signer.AccountNumber,
		Sequence:            sequence,
		IncludePQCPublicKey: opts.IncludePQCPublicKey,
	})
	if err != nil {
		return nil, err
	}
	res, err := c.broadcastBuilt(built)
	if err != nil {
		return nil, err
	}
	return &MigrateToHybridResult{
		AlreadyRegistered: ensure.AlreadyRegistered,
		RegisterTxHash:    ensure.TxHash,
		Result:            res,
	}, nil
}

// build signs the given messages into a BuiltTx using the helper's Signer
// (classical secp256k1 signature).
func (c *Client) build(msgs []sdk.Msg) (*tx.BuiltTx, error) {
	return tx.SendMessages(tx.SendMessagesParams{
		Account:       c.signer.Account,
		Messages:      msgs,
		Fee:           c.signer.Fee,
		ChainID:       c.signer.ChainID,
		AccountNumber: c.signer.AccountNumber,
		Sequence:      c.signer.Sequence,
	})
}

// broadcast builds, signs (classical), and broadcasts the given messages, waiting
// for inclusion.
func (c *Client) broadcast(msgs []sdk.Msg) (*tx.TxResult, error) {
	built, err := c.build(msgs)
	if err != nil {
		return nil, err
	}
	return c.broadcastBuilt(built)
}

// broadcastBuilt broadcasts an already-built tx and waits for inclusion.
func (c *Client) broadcastBuilt(built *tx.BuiltTx) (*tx.TxResult, error) {
	mode := c.signer.Mode
	if mode == "" {
		mode = tx.BroadcastSync
	}
	return tx.BroadcastAndWait(tx.BroadcastAndWaitParams{
		RestURL: c.signer.RestURL,
		TxBytes: built.TxRawBytes,
		Mode:    mode,
		Wait:    c.signer.Wait,
	})
}

File: packages/py/src/qorsdk/pqc_dx.py
"""Quantum-safe developer-experience (DX) helpers for QoreChain's ``x/pqc`` module.

QoreChain is **quantum-safe by default**: an account can register an ML-DSA-87
(Dilithium-5) post-quantum key, after which its transactions carry a hybrid
signature (classical secp256k1 + PQC) that the chain's ante handler verifies
end-to-end. These helpers make that migration a one-liner instead of
hand-rolling key-registration messages and status queries.

Status is read via the ``qor_getPQCKeyStatus`` JSON-RPC method (exposed as
``get_pqc_key_status`` on :class:`~qorsdk.qor.QorClient` /
:class:`~qorsdk.qor.AsyncQorClient`), whose result is shaped::

    {
        "address": "qor1...",
        "has_pqc_key": true,
        "key_type": "dilithium5",          # omitted when not registered
        "created_at_height": 1234           # omitted when not registered
    }

Registration is done with ``MsgRegisterPQCKeyV2``
(``/qorechain.pqc.v1.MsgRegisterPQCKeyV2``, the chain's classical-exempt
bootstrap path), which carries the signer's ML-DSA-87 public key with an
explicit ``algorithm_id`` plus its classical ECDSA (secp256k1) public key. Once
registered, send transactions through the hybrid signing path
(:func:`qorsdk.tx.build_hybrid_tx`) so every tx is quantum-safe.

This helper mirrors the ergonomic cross-VM / rollup helpers: it wraps the typed
``msg.pqc.register_pqc_key_v2`` / ``msg.pqc.migrate_pqc_key`` composers and the
SDK's tx surface (:func:`~qorsdk.tx.send_messages` /
:func:`~qorsdk.tx.build_hybrid_tx` + :func:`~qorsdk.tx.broadcast`) so an app can
check status and register/migrate keys without touching raw messages.
"""

from __future__ import annotations

from typing import Any, TypedDict

from .accounts import Secp256k1Account
from .messages import msg
from .messages._composer import Msg
from .pqc import ALGORITHM_DILITHIUM5, PqcKeypair
from .tx import broadcast, build_hybrid_tx, send_messages

#: The default ``key_type`` registered for an ML-DSA-87 (Dilithium-5) key,
#: matching the chain's ``x/pqc`` account ``KeyType`` string for hybrid
#: (classical + PQC) accounts.
HYBRID_KEY_TYPE = "hybrid"

#: Deprecated alias kept for backwards compatibility; the chain's bootstrap
#: registration path expects ``"hybrid"``.
DILITHIUM5_KEY_TYPE = HYBRID_KEY_TYPE

#: A Cosmos ``StdFee``-shaped dict (as produced by :func:`qorsdk.fees.estimate_fee`).
FeeDict = dict[str, Any]


class PqcStatus(TypedDict, total=False):
    """The resolved PQC registration status for an address.

    ``registered`` is always present; ``key_type``, ``algorithm_id`` and
    ``created_at_height`` are populated only when a key is registered.
    """

    registered: bool
    key_type: str
    algorithm_id: int
    created_at_height: int


class EnsureResult(TypedDict, total=False):
    """Result of :func:`ensure_pqc_registered`.

    ``already_registered`` is always present. ``tx_hash`` is set only when a
    ``MsgRegisterPQCKeyV2`` was broadcast (i.e. the key was missing before).
    """

    already_registered: bool
    tx_hash: str


def _algorithm_id_for_key_type(key_type: str) -> int | None:
    """Map a chain ``key_type`` string to a PQC algorithm id, if known."""
    if key_type.lower() in ("dilithium5", "ml-dsa-87", "ml_dsa_87"):
        return ALGORITHM_DILITHIUM5
    return None


def _resolve_status(raw: Any, address: str) -> PqcStatus:
    """Normalise a ``qor_getPQCKeyStatus`` result into a :class:`PqcStatus`."""
    data: dict[str, Any] = raw if isinstance(raw, dict) else {}
    registered = bool(data.get("has_pqc_key", False))
    status: PqcStatus = {"registered": registered}
    if registered:
        key_type = data.get("key_type")
        if key_type:
            status["key_type"] = key_type
            algorithm_id = _algorithm_id_for_key_type(key_type)
            if algorithm_id is not None:
                status["algorithm_id"] = algorithm_id
        created = data.get("created_at_height")
        if created is not None:
            status["created_at_height"] = int(created)
    return status


def get_pqc_status(client: Any, address: str) -> PqcStatus:
    """Read the PQC registration status for ``address``.

    ``client`` is a :class:`~qorsdk.qor.QorClient` (anything exposing
    ``get_pqc_key_status(address)``). Returns a :class:`PqcStatus` with
    ``registered`` plus, when registered, ``key_type`` / ``algorithm_id`` /
    ``created_at_height``.
    """
    raw = client.get_pqc_key_status(address)
    return _resolve_status(raw, address)


def is_pqc_registered(client: Any, address: str) -> bool:
    """True if ``address`` has a PQC key registered (via ``qor_getPQCKeyStatus``)."""
    return get_pqc_status(client, address)["registered"]


async def get_pqc_status_async(client: Any, address: str) -> PqcStatus:
    """Async variant of :func:`get_pqc_status` (for :class:`~qorsdk.qor.AsyncQorClient`)."""
    raw = await client.get_pqc_key_status(address)
    return _resolve_status(raw, address)


async def is_pqc_registered_async(client: Any, address: str) -> bool:
    """Async variant of :func:`is_pqc_registered`."""
    return (await get_pqc_status_async(client, address))["registered"]


def build_register_pqc_key(
    *,
    sender: str,
    dilithium_pubkey: bytes,
    ecdsa_pubkey: bytes,
    key_type: str = HYBRID_KEY_TYPE,
) -> Msg:
    """Build a ``MsgRegisterPQCKeyV2`` :class:`~qorsdk.messages.Msg` (no signing).

    Uses ``/qorechain.pqc.v1.MsgRegisterPQCKeyV2`` — the chain's current
    (classical-exempt bootstrap) registration path — with an explicit
    ``algorithm_id`` (ML-DSA-87 / Dilithium-5).
    """
    message: Msg = msg.pqc.register_pqc_key_v2(
        sender=sender,
        public_key=dilithium_pubkey,
        algorithm_id=ALGORITHM_DILITHIUM5,
        ecdsa_pubkey=ecdsa_pubkey,
        key_type=key_type,
    )
    return message


def build_migrate_pqc_key(
    *,
    sender: str,
    old_public_key: bytes,
    new_public_key: bytes,
    new_algorithm_id: int,
    old_signature: bytes,
    new_signature: bytes,
) -> Msg:
    """Build a ``MsgMigratePQCKey`` :class:`~qorsdk.messages.Msg` (no signing).

    Rotates an existing PQC key to ``new_public_key`` (under ``new_algorithm_id``).
    ``old_signature`` / ``new_signature`` prove control of both keys, as required
    by the chain's ``x/pqc`` migration handler.
    """
    message: Msg = msg.pqc.migrate_pqc_key(
        sender=sender,
        old_public_key=old_public_key,
        new_public_key=new_public_key,
        new_algorithm_id=new_algorithm_id,
        old_signature=old_signature,
        new_signature=new_signature,
    )
    return message


def _broadcast_messages(
    *,
    account: Secp256k1Account,
    messages: list[Msg],
    chain_id: str,
    account_number: int,
    sequence: int,
    fee: FeeDict,
    rest_url: str,
    pqc_keypair: PqcKeypair | None,
    memo: str,
    mode: str,
) -> Any:
    """Sign (hybrid when ``pqc_keypair`` given, else classical) and broadcast."""
    if pqc_keypair is not None:
        built = build_hybrid_tx(
            account=account,
            pqc_keypair=pqc_keypair,
            messages=messages,
            fee=fee,
            chain_id=chain_id,
            account_number=account_number,
            sequence=sequence,
            memo=memo,
        )
    else:
        built = send_messages(
            account=account,
            messages=messages,
            chain_id=chain_id,
            account_number=account_number,
            sequence=sequence,
            fee=fee,
            memo=memo,
        )
    return broadcast(rest_url, built.tx_raw_bytes, mode=mode)  # type: ignore[arg-type]


def _extract_tx_hash(response: Any) -> str | None:
    """Pull ``txhash`` out of a broadcast response, if present."""
    if isinstance(response, dict):
        tx_response = response.get("tx_response")
        if isinstance(tx_response, dict) and tx_response.get("txhash"):
            return str(tx_response["txhash"])
        if response.get("txhash"):
            return str(response["txhash"])
    return None


def ensure_pqc_registered(
    *,
    account: Secp256k1Account,
    pqc_keypair: PqcKeypair,
    chain_id: str,
    account_number: int,
    rest_url: str,
    fee: FeeDict,
    sequence: int = 0,
    qor: Any,
    key_type: str = HYBRID_KEY_TYPE,
    memo: str = "",
    mode: str = "sync",
) -> EnsureResult:
    """Make sure ``account`` has a PQC key registered; register it if missing.

    Idempotent: queries ``qor_getPQCKeyStatus`` first via ``qor`` (a
    :class:`~qorsdk.qor.QorClient`). If already registered, returns
    ``{"already_registered": True}`` with no transaction. Otherwise builds and
    broadcasts a ``MsgRegisterPQCKeyV2`` carrying the signer's Dilithium public key
    (``pqc_keypair.public_key``) and classical ECDSA public key
    (``account.public_key``), returning
    ``{"already_registered": False, "tx_hash": ...}``.

    The registration tx itself is signed classically (a wallet cannot yet sign
    hybrid for the very key it is registering); subsequent txs should use the
    hybrid path — see :func:`migrate_to_hybrid`.
    """
    if is_pqc_registered(qor, account.address):
        return {"already_registered": True}

    register = build_register_pqc_key(
        sender=account.address,
        dilithium_pubkey=pqc_keypair.public_key,
        ecdsa_pubkey=account.public_key,
        key_type=key_type,
    )
    response = _broadcast_messages(
        account=account,
        messages=[register],
        chain_id=chain_id,
        account_number=account_number,
        sequence=sequence,
        fee=fee,
        rest_url=rest_url,
        pqc_keypair=None,
        memo=memo,
        mode=mode,
    )
    result: EnsureResult = {"already_registered": False}
    tx_hash = _extract_tx_hash(response)
    if tx_hash is not None:
        result["tx_hash"] = tx_hash
    return result


def migrate_to_hybrid(
    *,
    account: Secp256k1Account,
    pqc_keypair: PqcKeypair,
    messages: list[Msg],
    chain_id: str,
    account_number: int,
    rest_url: str,
    fee: FeeDict,
    sequence: int = 0,
    qor: Any,
    key_type: str = HYBRID_KEY_TYPE,
    register_sequence: int | None = None,
    memo: str = "",
    mode: str = "sync",
) -> Any:
    """Ensure ``account`` is PQC-registered, then send ``messages`` quantum-safe.

    This is the "go quantum-safe" path: it calls :func:`ensure_pqc_registered`
    (broadcasting a ``MsgRegisterPQCKeyV2`` if needed, signed classically at
    ``register_sequence`` — defaulting to ``sequence``), then signs and
    broadcasts ``messages`` through the hybrid path
    (:func:`qorsdk.tx.build_hybrid_tx`) so the transaction carries both a
    classical and an ML-DSA-87 signature.

    Returns the decoded broadcast response for the hybrid ``messages`` tx. If a
    registration tx was broadcast first, advance ``sequence`` accordingly (pass
    the registration at ``register_sequence`` and the hybrid tx at ``sequence``).

    :raises ValueError: If ``messages`` is empty.
    """
    if not messages:
        raise ValueError("migrate_to_hybrid requires at least one message to send")

    ensure_pqc_registered(
        account=account,
        pqc_keypair=pqc_keypair,
        chain_id=chain_id,
        account_number=account_number,
        rest_url=rest_url,
        fee=fee,
        sequence=sequence if register_sequence is None else register_sequence,
        qor=qor,
        key_type=key_type,
        memo=memo,
        mode=mode,
    )

    return _broadcast_messages(
        account=account,
        messages=messages,
        chain_id=chain_id,
        account_number=account_number,
        sequence=sequence,
        fee=fee,
        rest_url=rest_url,
        pqc_keypair=pqc_keypair,
        memo=memo,
        mode=mode,
    )


def migrate_pqc_key(
    *,
    account: Secp256k1Account,
    old_public_key: bytes,
    new_public_key: bytes,
    new_algorithm_id: int,
    old_signature: bytes,
    new_signature: bytes,
    chain_id: str,
    account_number: int,
    rest_url: str,
    fee: FeeDict,
    sequence: int = 0,
    pqc_keypair: PqcKeypair | None = None,
    memo: str = "",
    mode: str = "sync",
) -> Any:
    """Build, sign, and broadcast a ``MsgMigratePQCKey`` (rotate the PQC key).

    Rotates ``old_public_key`` to ``new_public_key`` under ``new_algorithm_id``,
    with ``old_signature`` / ``new_signature`` proving control of both keys. The
    tx is hybrid-signed when ``pqc_keypair`` is supplied, else classical.
    Returns the decoded broadcast response.
    """
    migrate = build_migrate_pqc_key(
        sender=account.address,
        old_public_key=old_public_key,
        new_public_key=new_public_key,
        new_algorithm_id=new_algorithm_id,
        old_signature=old_signature,
        new_signature=new_signature,
    )
    return _broadcast_messages(
        account=account,
        messages=[migrate],
        chain_id=chain_id,
        account_number=account_number,
        sequence=sequence,
        fee=fee,
        rest_url=rest_url,
        pqc_keypair=pqc_keypair,
        memo=memo,
        mode=mode,
    )


__all__ = [
    "HYBRID_KEY_TYPE",
    "DILITHIUM5_KEY_TYPE",
    "PqcStatus",
    "EnsureResult",
    "get_pqc_status",
    "is_pqc_registered",
    "get_pqc_status_async",
    "is_pqc_registered_async",
    "build_register_pqc_key",
    "build_migrate_pqc_key",
    "ensure_pqc_registered",
    "migrate_to_hybrid",
    "migrate_pqc_key",
]
